File: idg/cpp/src/IDG.cpp
#include <complex>
#include <cstddef>

#include <omp.h>

#include <xtensor-fftw/basic.hpp>
#include <xtensor-fftw/helper.hpp>
#include <xtensor/containers/xarray.hpp>
#include <xtensor/core/xtensor_forward.hpp>
#include <xtensor/views/xview.hpp>

#include "IDG.h"
#include "idgtypes.h"
#include "kernels.h"

void Gridder::grid_onto_subgrids(
    float w_step, float image_size, size_t grid_size,
    const xt::xarray<float> &wavenumbers, const xt::xarray<UVW> &uvw,
    const xt::xarray<std::complex<float>> &visibilities,
    const xt::xarray<float> &taper, const xt::xarray<Metadata> &metadata,
    xt::xarray<std::complex<float>> &subgrids) const {
  assert(nr_correlations_in_ == visibilities.shape()[3]);
  assert(nr_correlations_out_ == subgrids.shape()[1]);
  assert(subgrid_size_ == subgrids.shape()[2]);
  const size_t nr_subgrids = metadata.size();

#pragma omp parallel for schedule(dynamic)
  for (size_t s = 0; s < nr_subgrids; ++s) {
    auto subgrid =
        xt::eval(xt::view(subgrids, s, xt::all(), xt::all(), xt::all()));

    visibilities_to_subgrid(s, metadata, w_step, grid_size, image_size,
                            wavenumbers, visibilities, uvw, taper,
                            nr_correlations_in_, subgrid_size_, subgrid);

    xt::view(subgrids, s, xt::all(), xt::all(), xt::all()) = subgrid;
  }
}

void Gridder::ifft_subgrids(xt::xarray<std::complex<float>> &subgrids) const {
  for (size_t s = 0; s < subgrids.shape(0); ++s) {
    for (size_t c = 0; c < subgrids.shape(1); ++c) {
      auto subgrid = xt::eval(xt::view(subgrids, s, c, xt::all(), xt::all()));

      subgrid = xt::fftw::ifft2(subgrid);

      xt::view(subgrids, s, c, xt::all(), xt::all()) = subgrid;
    }
  }
}

void Gridder::add_subgrids_to_grid(
    const xt::xarray<Metadata> &metadata,
    const xt::xarray<std::complex<float>> &subgrids,
    xt::xarray<std::complex<float>> &grid) const {
  const size_t nr_correlations = grid.shape()[0];
  const size_t grid_size = grid.shape()[1];
  const size_t nr_subgrids = subgrids.shape()[0];
  const size_t subgrid_size = subgrids.shape()[2];

  xt::xarray<std::complex<float>> phasor =
      compute_phasor(static_cast<int>(subgrid_size));

  for (size_t s = 0; s < nr_subgrids; ++s) {
    add_subgrid_to_grid(s, metadata, subgrids, grid, phasor, nr_correlations,
                        subgrid_size, grid_size);
  }
}

void Gridder::transform(xt::xarray<std::complex<float>> &grid) const {
  assert(nr_correlations_out_ == grid.shape()[0]);
  const size_t height = grid.shape()[1];
  const size_t width = grid.shape()[2];
  assert(height == width);

  for (size_t i = 0; i < nr_correlations_out_; ++i) {
    auto slice = xt::view(grid, i, xt::all(), xt::all());
    auto tmp = xt::xarray<std::complex<float>>::from_shape(slice.shape());

    for (size_t x = 0; x < width; ++x) {
      for (size_t y = 0; y < height; ++y) {
        size_t dst_x = (x + width / 2) % width;
        size_t dst_y = (y + height / 2) % height;
        tmp(dst_y, dst_x) = slice(y, x);
      }
    }

    tmp = xt::fftw::ifft2(tmp);

    for (size_t x = 0; x < width; ++x) {
      for (size_t y = 0; y < height; ++y) {
        size_t src_x = (x + width / 2) % width;
        size_t src_y = (y + height / 2) % height;
        slice(y, x) = tmp(src_y, src_x);
      }
    }

    std::complex<float> scale{2.0f, 0.0f};

    slice *= scale;

    xt::view(grid, i, xt::all(), xt::all()) = slice;
  }
}

File: idg/python/idg.py
from concurrent.futures import ThreadPoolExecutor, as_completed
import multiprocessing
import numpy as np

import idgtypes
from kernels import add_subgrid_to_grid, compute_phasor, visibilities_to_subgrid


class Gridder:

    def __init__(self, nr_correlations_in: int, subgrid_size: int):
        self.nr_correlations_in = nr_correlations_in
        self.nr_correlations_out = 4 if nr_correlations_in == 4 else 1
        self.subgrid_size = subgrid_size

    def grid_onto_subgrids(
        self,
        w_step: float,
        image_size: float,
        grid_size: int,
        wavenumbers: np.ndarray,
        uvw: np.ndarray,
        visibilities: np.ndarray,
        taper: np.ndarray,
        metadata: np.ndarray,
        subgrids: np.ndarray,
    ) -> None:
        """
        Grid visibilities onto subgrids.

        :param w_step: w step in wavelengths
        :param image_size: image size in radians
        :param grid_size: size of the grid
        :param wavenumbers: array of wavenumbers, shape (nr_channels)
        :param uvw: array of uvw coordinates, shape (nr_baselines, nr_time, 3)
        :param visibilities: array of visibilities, shape (nr_baselines, nr_time, nr_channels, nr_correlations_in)
        :param taper: array of the taper function, shape (subgrid_size, subgrid_size)
        :param metadata: array with metadata, shape (nr_subgrids)
        :param subgrids: array of subgrids, shape (nr_subgrids, nr_correlations_out, subgrid_size, subgrid_size)
        """
        assert self.nr_correlations_in == visibilities.shape[3]
        assert self.nr_correlations_out == subgrids.shape[1]
        assert self.subgrid_size == subgrids.shape[2]
        nr_subgrids = metadata.shape[0]

        # Grid visibilities onto subgrids
        with ThreadPoolExecutor(max_workers=multiprocessing.cpu_count()) as executor:
            futures = []
            for s in range(nr_subgrids):
                future = executor.submit(
                    visibilities_to_subgrid,
                    s,
                    metadata,
                    w_step,
                    grid_size,
                    image_size,
                    wavenumbers,
                    visibilities,
                    uvw,
                    taper,
                    self.nr_correlations_in,
                    self.subgrid_size,
                    subgrids[s],
                )
                futures.append(future)

            for future in as_completed(futures):
                pass

        # Apply FFT to each subgrid
        subgrids[:] = np.fft.ifft2(subgrids, axes=(2, 3))

    def add_subgrids_to_grid(
        self,
        metadata: np.ndarray,  # metadata array
        subgrids: np.ndarray,  # subgrid array
        grid: np.ndarray,  # grid array
    ) -> None:
        """
        Add subgrids to the grid.

        :param metadata: metadata arraym shape (nr_subgrids)
        :param subgrids: subgrid array, shape (nr_subgrids, nr_correlations_out, subgrid_size, subgrid_size)
        :param grid: grid array (nr_correlations_out, grid_size, grid_size)
        """

        nr_correlations = grid.shape[0]
        grid_size = grid.shape[1]
        nr_subgrids = subgrids.shape[0]
        subgrid_size = subgrids.shape[2]

        # Compute phasor
        phasor = compute_phasor(subgrid_size)

        # Iterate all subgrids
        for s in range(nr_subgrids):
            add_subgrid_to_grid(
                s,
                metadata,
                subgrids,
                grid,
                phasor,
                nr_correlations,
                subgrid_size,
                grid_size,
            )

    def transform(self, direction: int, grid: np.ndarray) -> None:
        """
        Transform Fourier Domain<->Image Domain.

        :param direction: idg.FourierDomainToImageDomain or idg.ImageDomainToFourierDomain
        :param grid array, shape (nr_correlations_out, grid_size, grid_size)
        """
        assert self.nr_correlations_out == grid.shape[0]
        height = grid.shape[1]
        width = grid.shape[2]
        assert height == width

        # FFT shift
        grid[:] = np.fft.fftshift(grid, axes=(1, 2))

        # FFT
        if direction == idgtypes.FOURIER_DOMAIN_TO_IMAGE_DOMAIN:
            grid[:] = np.fft.ifft2(grid, axes=(1, 2))
        else:
            grid[:] = np.fft.fft2(grid, axes=(1, 2))

        # FFT shift
        grid[:] = np.fft.fftshift(grid, axes=(1, 2))

        # Scaling
        scale = 2 + 0j
        if direction == idgtypes.FOURIER_DOMAIN_TO_IMAGE_DOMAIN:
            grid *= scale
        else:
            grid /= scale
